### PR TITLE
 Add exposedmethod decorator and documentation (Resolves #292)

### LINKED ISSUE #292: recommend use of a decorator rather than expose_*
rpyc: absolutely awesome.

recommendation: decorator that exposes properties explicitly.  this was a technique used in the pyjs jsonrpc service (which hilariously was 20 lines... gosh shock gasp 50 if you wanted remote error handling)

### PR DESCRIPTION
**Pull request by comrumino**

### CODE REVIEW

**Comment by coldfix** [comment]
Hi, I don't think patching `Server` is the correct approach because services are not only exposed by servers, and because this means that you modify the class later than it was constructed, potentially many times. I would prefer to either

- provide a class decorator that patches the class directly
- use a metaclass for Service

(even if both of these methods are not directly compatible with a runtime-dependent `exposed_prefix`)

**Comment by comrumino** [comment]
Thanks for the feedback. I managed to keep support for the runtime-dependent `exposed_prefix` :smiley:.

**Comment by coldfix** [comment]
Hey, I haven't forgotten about this PR, I will take some time to review. Another point is that I would prefer to have policy proxies (*) in the long run, if possible compatible with the notation introduced in this PR. Maybe the class decorator should be renamed to something like `rpyc.with_policy`. Thinking further, the policy wrapper object won't even need a class decorator and look much cleaner from a user perspective. I'll play around with policy wrappers and get back to you in a week or so.

(*) My basic idea is that `rpyc.Connection` won't access the "real" object directly anymore, but will interact with it through a proxy that basically does all the `handle_XXX` stuff in the rpyc protocol, i.e. getattr/setattr/delattr/operators/... This proxy would be free to enforce arbitrary custom access policies on a very granular level, i.e. you can have different policies for different objects of the same type in the same connection, or even different views of the same object. There are two approaches that will have to be evaluated: a *policy* with `handle_XXX` methods versus seemless *wrapper* objects that behave exactly the same as the target by implementing the corresponding dunder methods while `handle_XXX` stays in `Connection`. (The former will typically need instanciating less proxies, while the latter allows to "cut the middle man" and access objects directly and makes it easier to have different policy views on the same object)

**Comment by comrumino** [comment]
Agreed, the name `rpyc.service` isn't great. Hiding the class decorator function would be a better API decision (but definitely requires some overhauling elsewhere to work). Maybe the destination of this PR should be a different branch? I'd be happy to change the destination if you would like.

**Comment by comrumino** [comment]
> Nevermind all the micro-comments, I think the solution should do exactly the same as suggested in:
> 
> [#70 (comment)](https://github.com/tomerfiliba/rpyc/issues/70#issuecomment-4640625)
> 
> With one exception: replace the `delattr(cls, k)` line with another `setattr(cls, k, v.func)`.

The current commit should behave exactly as desired, but it also supports exposed classes that have exposed methods and any non-standard exposed_prefix (which is why it's a bit longer)

File: rpyc/utils/__init__.py [hunk 1]
```diff
@@ -1,4 +1,55 @@
...
 
+
+try:
+    # python3 depricated basestring, so if this causes exception catch and alias to str
+    basestring
+except Exception:
+    basestring = str
+
+
+def service(cls):
```
[the remark this answers is not in the document]
Comment (comrumino) [reply rpyc/utils/__init__.py]: For doing something like `rpyc.with_policy('policy_name')` 
```
def _policy_wrapper(policy_name, cls):
    # lookup policy name perform
    # apply policy to cls
    return service(cls)

def policy(arg):
    if isinstance(arg, basestring):
        return functools.partial(_policy_wrapper, arg)
    elif hasattr(arg, '__call__'):
        return _policy_wrapper(default_policy_name, arg)
    else:
        raise TypeError('rpyc.policy expects a callable object or a string')

```
I think thats what you getting at?

File: rpyc/utils/__init__.py [hunk 2]
```diff
@@ -1,4 +1,55 @@
...
+            raise ValueError(_err)
+        attr_obj.__name__ = renamed
+        if attr_obj.__class__.__name__ != 'function':  # recurse exposed objects such as a class
+            attr_obj = service(attr_obj)
+        delattr(cls, name)
+        setattr(cls, renamed, attr_obj)
+    return cls
+
+
+def exposed(arg):
```
[the remark this answers is not in the document]
Comment (comrumino) [reply rpyc/utils/__init__.py]: I'm okay with this if we want to drop support for non-stander prefixes.

File: rpyc/utils/__init__.py [hunk 3]
```diff
@@ -1,4 +1,55 @@
...
+    """find and rename exposed decorated attributes"""
+    decorated_attrs = []
+    for attr_obj in cls.__dict__.values():  # find exposed decorated attributes
+        if hasattr(attr_obj, '__exposed__') and attr_obj.__exposed__:
+            decorated_attrs.append(attr_obj)
+    for attr_obj in decorated_attrs:  # rebind exposed decorated attributes
+        exposed_prefix = attr_obj.__exposed__
+        name = attr_obj.__name__
+        renamed = exposed_prefix + name
+        if hasattr(cls, renamed):
```
[the remark this answers is not in the document]
Comment (comrumino) [reply rpyc/utils/__init__.py]: setting to false after creating the alias seems to be problematic

**Pull request closed without merging**